File: python/algo/202508/algo0662.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# 二叉树 - 2.13 二叉树 BFS
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_width = 1
        # 存储当前层节点和对应的索引
        current_nodes = [root]
        current_indices = [0]
        
        while current_nodes:
            # 计算当前层宽度（只有当节点数大于1时才需要更新）
            if len(current_indices) > 1:
                current_width = current_indices[-1] - current_indices[0] + 1
                if current_width > max_width:
                    max_width = current_width
            
            # 准备下一层的节点和索引
            next_nodes = []
            next_indices = []
            
            # 遍历当前层节点，记录下一层节点及其索引
            for i in range(len(current_nodes)):
                node = current_nodes[i]
                idx = current_indices[i]
                
                # 左子节点索引：2*idx + 1
                if node.left:
                    next_nodes.append(node.left)
                    next_indices.append(idx * 2 + 1)
                
                # 右子节点索引：2*idx + 2
                if node.right:
                    next_nodes.append(node.right)
                    next_indices.append(idx * 2 + 2)
            
            # 移动到下一层
            current_nodes = next_nodes
            current_indices = next_indices
        
        return max_width

File: cplus/leetcode/202508/algo0662.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// 二叉树 - 2.13 二叉树 BFS
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        // 自行解答 - 内存溢出
        int mx_widht = 1;
        deque<TreeNode*> cur; cur.push_back(root);
        while (!cur.empty()) {
            mx_widht = max(mx_widht, (int)cur.size());
            deque<TreeNode*> nxt;
            for (auto& node:cur) {
                if (node) {
                    nxt.push_back(node->left); nxt.push_back(node->right);
                }else {
                    nxt.push_back(nullptr); nxt.push_back(nullptr);
                }
            }
            while (!nxt.empty() && !nxt.front()) nxt.pop_front();
            while (!nxt.empty() && !nxt.back()) nxt.pop_back();
            cur = move(nxt);
        }

        return mx_widht;
    }

    int widthOfBinaryTree2(TreeNode* root) {
        // 自行解答
        unsigned long long mx_widht = 1;
        vector<unsigned long long> inx; inx.push_back(0);
        vector<TreeNode*>cur; cur.push_back(root);
        while (!cur.empty()) {
            if (inx.size()>1) mx_widht = max(mx_widht, inx.back()-inx[0]+1);
            vector<TreeNode*> nxt; vector<unsigned long long> n_inx;
            int n = cur.size();
            for (int i=0; i<n; ++i) {
                if (cur[i]->left) {
                    nxt.push_back(cur[i]->left); n_inx.push_back(inx[i]*2+1);
                }
                if (cur[i]->right) {
                    nxt.push_back(cur[i]->right); n_inx.push_back(inx[i]*2+2);
                }                    
            }
            cur = move(nxt); inx = move(n_inx);
        }

        return mx_widht;
    }

};